common/cuda_hip/factorization/cholesky_kernels.hpp.inc

```diff
@@ -156,3 +144,272 @@ __launch_bounds__(default_block_size) void cholesky_symbolic_factorize_kernel(
+    matrix::Csr<ValueType, IndexType>* l_factor,
+    const array<IndexType>& tmp_storage)
+{
+    const auto num_rows = static_cast<IndexType>(mtx->get_size()[0]);
+    if (num_rows == 0) {
+        return;
+    }
+    const auto mtx_nnz = static_cast<IndexType>(mtx->get_num_stored_elements());
+    const auto postorder_cols = tmp_storage.get_const_data();
+    const auto lower_ends = postorder_cols + mtx_nnz;
+    const auto row_ptrs = mtx->get_const_row_ptrs();
+    const auto postorder = forest.postorder.get_const_data();
+    const auto inv_postorder = forest.inv_postorder.get_const_data();
+    const auto postorder_parent = forest.postorder_parents.get_const_data();
+    const auto out_row_ptrs = l_factor->get_const_row_ptrs();
+    const auto out_cols = l_factor->get_col_idxs();
+    const auto num_blocks =
+        ceildiv(num_rows, default_block_size / config::warp_size);
+    kernel::symbolic_factorize<config::warp_size>
+        <<<num_blocks, default_block_size>>>(
+            num_rows, row_ptrs, lower_ends, inv_postorder, postorder_cols,
+            postorder, postorder_parent, out_row_ptrs, out_cols);
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_CHOLESKY_SYMBOLIC_FACTORIZE);
+
+
+template <typename IndexType>
+void build_children_from_parents(
+    std::shared_ptr<const DefaultExecutor> exec,
+    gko::factorization::elimination_forest<IndexType>& forest)
+{
+    const auto num_rows = forest.parents.get_num_elems();
+    // build COO representation of the tree
+    array<IndexType> col_idx_array{exec, num_rows};
+    const auto col_idxs = col_idx_array.get_data();
+    const auto parents = forest.parents.get_const_data();
+    const auto children = forest.children.get_data();
+    const auto child_ptrs = forest.child_ptrs.get_data();
+    exec->copy(num_rows, parents, col_idxs);
+    thrust::sequence(thrust::device, children, children + num_rows,
+                     IndexType{});
+    // group by parent
+    thrust::stable_sort_by_key(thrust::device, col_idxs, col_idxs + num_rows,
+                               children);
+    // create child pointers for groups of children
+    components::convert_idxs_to_ptrs(exec, col_idxs, num_rows,
+                                     num_rows + 1,  // rows plus sentinel
+                                     child_ptrs);
+}
+
+
+template <typename IndexType>
+__global__ void compute_subtree_sizes_kernel(
+    const IndexType* __restrict__ parents,
+    const IndexType* __restrict__ child_ptrs,
+    const IndexType* __restrict__ children, IndexType* __restrict__ postorder,
+    size_type num_rows)
+{
+    const auto tidx = thread::get_thread_id_flat();
+    __shared__ int ready[default_block_size];
+    ready[tidx] = 0;
+    __syncthreads();
+    if (tidx >= num_rows) {
+        return;
+    }
+
+    const auto child_begin = child_ptrs[num_rows];
+    const auto child_end = child_ptrs[num_rows + 1];
+    for (auto i = child_begin; i < child_end; i++) {
```

Comment:
What is the purpose of this loop? Is something missing?

Reply:
oh right, I left some WIP code for postordering the tree on the GPU. Let me remove that, I can always recover it from the history